KAFKA-14982: Improve the kafka-metadata-quorum output

fvaleri opened the pull request:
When running kafka-metadata-quorum script to get the quorum replication status, the LastFetchTimestamp and LastCaughtUpTimestamp output is not human-readable.

I will be convenient to add an optional flag (-hr, --human-readable) to enable a human-readable format showing the delay in ms (i.e. 366 ms ago).

This dealy is computed as (now - timestamp), where they are both represented as Unix time (UTC based).

```sh
$ bin/kafka-metadata-quorum.sh --bootstrap-server :9092 describe --replication --human-readable
NodeId	LogEndOffset	Lag	LastFetchTimestamp	LastCaughtUpTimestamp	Status  	
2     	61          	0  	5 ms ago          	5 ms ago             	Leader  	
3     	61          	0  	56 ms ago         	56 ms ago            	Follower	
4     	61          	0  	56 ms ago         	56 ms ago            	Follower
```
Implementation of [KIP-927](https://cwiki.apache.org/confluence/display/KAFKA/KIP-927%3A+Improve+the+kafka-metadata-quorum+output).

fvaleri committed:
KAFKA-14982: Improve the kafka-metadata-quorum output

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

machi1990 commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -155,17 +163,28 @@ private static void handleDescribeReplication(Admin admin) throws ExecutionExcep
+    private static List<List<String>> quorumInfoToRows(QuorumInfo.ReplicaState leader,
+                                                       Stream<QuorumInfo.ReplicaState> infos,
+                                                       String status,
+                                                       boolean readable) {
```
```suggestion
                                                       boolean humanReadable) {
```
nit: suggestion to change `readable` to `humanReadable`, if accepted, it has to be applied throughout: the caller and the callee of this method.

machi1990 commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -138,15 +146,15 @@ private static void addDescribeParser(Subparsers subparsers) {
     }
 
-    private static void handleDescribeReplication(Admin admin) throws ExecutionException, InterruptedException {
+    private static void handleDescribeReplication(Admin admin, boolean readable) throws ExecutionException, InterruptedException {
```
same comment as https://github.com/apache/kafka/pull/13738/files#r1199737177

machi1990 reviewed (approved):
lgtm

I left one naming suggestion

divijvaidya commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -155,17 +163,28 @@ private static void handleDescribeReplication(Admin admin) throws ExecutionExcep
+                                                       String status,
+                                                       boolean readable) {
+        return infos.map(info -> {
+            final double nowMs = System.currentTimeMillis();
```
Consider using `Time.SYSTEM`

divijvaidya replied:
`System.currentTimeMillis()` returns a long. https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#currentTimeMillis

Using long return type here will makes subtraction easier later on in the code.

divijvaidya commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -155,17 +163,28 @@ private static void handleDescribeReplication(Admin admin) throws ExecutionExcep
+        return infos.map(info -> {
+            final double nowMs = System.currentTimeMillis();
+            String lastFetchTimestamp = !info.lastFetchTimestamp().isPresent() ? "-1" :
+                valueOf(readable ? format("%.0f ms ago", nowMs - info.lastFetchTimestamp().getAsLong())
```
please expect skew/drift in system clock. For such situations we should handle the case when difference is negative.

divijvaidya commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -79,6 +82,10 @@ static void execute(String... args) throws Exception {
             .addArgument("--command-config")
             .type(Arguments.fileType())
             .help("Property file containing configs to be passed to Admin Client.");
+        parser
```
this should probably go inside `addDescribeParser` since we want to parse for this only when --describe is called

divijvaidya commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -92,11 +99,12 @@ static void execute(String... args) throws Exception {
             props.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, namespace.getString("bootstrap_server"));
             admin = Admin.create(props);
 
+            boolean readable = Optional.of(namespace.getBoolean("human_readable")).orElse(false);
```
we only use this for --describe and replication, please move inside `handleDescribeReplication()`

divijvaidya commented on tools/src/test/java/org/apache/kafka/tools/MetadataQuorumCommandTest.java:
```diff
@@ -172,4 +173,22 @@ public void testDescribeQuorumInZkMode() {
+    @ClusterTests({
+        @ClusterTest(clusterType = Type.CO_KRAFT, brokers = 1, controllers = 1),
+    })
+    public void testHumanReadableTimestamps() {
```
this test doesn't test the business logic that we introduced in the code, i.e. the part about ensuring X in "X ms ago" is what is desired is missing.

fvaleri committed:
Add Manyanda and Divij feedback

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

divijvaidya commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -136,17 +143,21 @@ private static void addDescribeParser(Subparsers subparsers) {
             .help("Detailed information about the status of replication")
             .action(Arguments.storeTrue());
+        replicationArgs
+            .addArgument("-hr", "--human-readable")
```
This doesn't work without `--replication`, right? In which case, this should probably be a subParser of replication (similar to how we handle subparsing for --describe)

fvaleri replied:
It doesn't work with --status, but I see what you mean. We should raise an error, rather than silently ignore the --hr option.

divijvaidya commented on tools/src/test/java/org/apache/kafka/tools/MetadataQuorumCommandTest.java:
```diff
@@ -172,4 +173,34 @@ public void testDescribeQuorumInZkMode() {
+        @ClusterTest(clusterType = Type.CO_KRAFT, brokers = 1, controllers = 1),
+    })
+    public void testHumanReadableTimestamps() {
+        String replicationOutput0 = ToolsTestUtils.captureStandardOut(() ->
```
please add a test to assert that `-hr` fails gracefully if used without `--replication`

fvaleri replied:
Yes, this is related to the addArgument comment. Will do.

divijvaidya reviewed (commented):
Thank for your making the changes. One last part left to fix.

fvaleri committed:
Fail gracefully when -hr is used without --replication

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

divijvaidya reviewed (approved):
Changes look good to me!

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -155,17 +169,35 @@ private static void handleDescribeReplication(Admin admin) throws ExecutionExcep
+    }
+
+    private static long durationMs(long timestampMs) {
+        Instant instant = Instant.ofEpochMilli(timestampMs);
```
This variable name is hard to understand, I don't know which name is better, maybe `timestampBefore`?

fvaleri replied:
Renamed to lastTimestamp.

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -155,17 +169,35 @@ private static void handleDescribeReplication(Admin admin) throws ExecutionExcep
+        Instant instant = Instant.ofEpochMilli(timestampMs);
+        Instant now = Instant.now();
+        if (!(instant.isAfter(Instant.EPOCH) && instant.isBefore(now))) {
+            throw new KafkaException("Invalid timestamp, possible drift in system clock");
```
Could we add `now` and `timestampMs` into the error message?

fvaleri replied:
Sure, I reworked that code to include more details.

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -155,17 +169,35 @@ private static void handleDescribeReplication(Admin admin) throws ExecutionExcep
+        }).collect(Collectors.toList());
+    }
+
+    private static long durationMs(long timestampMs) {
```
Could we add a test for this method? It looks like we only test the output format, not test the output duration is correct.

fvaleri replied:
Done.

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -131,22 +140,27 @@ private static void addDescribeParser(Subparsers subparsers) {
             .action(Arguments.storeTrue());
+        replicationArgs
+            .addArgument("-hr", "--human-readable")
+            .help("Print human-readable timestamps")
```
Since we're not printing timestamp with `-hr` option, maybe change to:
nit: Print human-readable timestamps duration.

fvaleri replied:
What about "Human-readable output"? It's short, clear and allows reuse. In the future and even in other tools, we could use this flag to make some fields readable or filter out less important details that make the output hard to read.

showuon reviewed (commented):
@fvaleri , thanks for the PR. Left some comments.

fvaleri committed:
Add feedback from Luke

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri commented:
@showuon all your comments should be addressed with the latest commit. 

I removed the `-hr` abbreviation and updated the KIP because argparse4j supports them by default. For example you can do the following:

```sh
$ bin/kafka-metadata-quorum.sh --bootstrap-server :9092 describe --re --hu
NodeId	LogEndOffset	Lag	LastFetchTimestamp	LastCaughtUpTimestamp	Status  	
2     	2032        	0  	3 ms ago          	3 ms ago             	Leader  	
3     	2032        	0  	207 ms ago        	207 ms ago           	Follower	
4     	2032        	0  	207 ms ago        	207 ms ago           	Follower
```

fvaleri commented:
Failing tests are unrelated.

showuon commented on tools/src/test/java/org/apache/kafka/tools/MetadataQuorumCommandErrorTest.java:
```diff
@@ -45,4 +51,14 @@ public void testDescribeOptions() {
     }
 
+    @Test
+    public void testDelayMs() {
+        long nowMs = Instant.now().toEpochMilli();
+        assertTrue(MetadataQuorumCommand.delayMs(nowMs, "test") >= 0);
+        long invalidEpochMs = Instant.EPOCH.minus(1, ChronoUnit.DAYS).toEpochMilli();
+        assertThrows(KafkaException.class, () -> MetadataQuorumCommand.delayMs(invalidEpochMs, "test"));
+        long futureTimestampMs = Instant.now().plus(1, ChronoUnit.DAYS).toEpochMilli();
+        assertThrows(KafkaException.class, () -> MetadataQuorumCommand.delayMs(futureTimestampMs, "test"));
+    }
```
Thanks for adding it. Looks good.

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -191,13 +191,16 @@ private static List<List<String>> quorumInfoToRows(QuorumInfo.ReplicaState leade
 
-    private static long durationMs(long timestampMs) {
-        Instant instant = Instant.ofEpochMilli(timestampMs);
+    /* test */ static long delayMs(long timestampMs, String desc) {
```
nit: In Kafka, we usually put the test comment on one line above, i.e.
```
// visible for testing
static long delayMs(long timestampMs, String desc) {
```

showuon replied:
Also, the method `delayMs` doesn't look right to me. I did a google search, it looks like the `relative time` is the correct name for it. Check [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat/format) and [here](https://momentjs.com/docs/#/displaying/fromnow/). So, maybe `relativeTimeMs`?

fvaleri replied:
Sure, I like relativeTimeMs. Thanks.

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -191,13 +191,16 @@ private static List<List<String>> quorumInfoToRows(QuorumInfo.ReplicaState leade
-    private static long durationMs(long timestampMs) {
-        Instant instant = Instant.ofEpochMilli(timestampMs);
+    /* test */ static long delayMs(long timestampMs, String desc) {
+        Instant lastTimestamp = Instant.ofEpochMilli(timestampMs);
```
`lastTimestamp` LGTM. Thanks.

showuon reviewed (commented):
@fvaleri , thanks for the update. Left some comments.

fvaleri committed:
Use relativeTimeMs and change test comment

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -96,14 +100,18 @@ static void execute(String... args) throws Exception {
+                    handleDescribeReplication(admin, humanReadable);
                 } else if (namespace.getBoolean("status")) {
+                    if (namespace.getBoolean("human_readable")) {
+                        throw new TerseException("The option -hr/--human-readable is only supported along with --replication");
```
Since we don't have `-hr` option, this message should be updated.

fvaleri replied:
Thanks for catching this.

showuon reviewed (commented):
One minor comment left. Thanks.

fvaleri committed:
Fix error message

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon reviewed (approved):
LGTM!

showuon commented:
Failed tests are unrelated:
```
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testReplication()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector
    Build / JDK 11 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor()
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.tools.MetadataQuorumCommandErrorTest.testRelativeTimeMs()
```

dajac commented:
@fvaleri testRelativeTimeMs seems to fail regularly. There is an example here: https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13782/1/tests. Could you take a look?
